File: src/GeneratorFlag.js
class GeneratorFlag(callbackToAppToAlterPosition) extends React.Component {

  class GeneratorFlagSegmentType {
    static NE = new GeneratorFlagSegmentType('NE');
    static E = new GeneratorFlagSegmentType('E');
    static SE = new GeneratorFlagSegmentType('SE');
  }

  const [selectedSegment, setSelectedSegment] = useState(null);
	const leftCoords = null; // the coords of the left corner of the flag. the other corners are inferred

  constructor(props) {
    super(props);
    this.state = {
      leftCoords: props.c
    };
  }

  // given a corner on this flag, 
  // returns [] of two more flags that will complete this intersection
  spawnAdjacentFlags(c) {
    const result = [];
    if (c.equals(getLeftCoords())) {
      result.push(new GeneratorFlag(c.posIncrW()));
      result.push(new GeneratorFlag(c.posIncrSW()));
    } else if (c.equals(getRightCoords())) {
      result.push(new GeneratorFlag(c);
      result.push(new GeneratorFlag(c.posIncrSW()));
    } else if (c.equals(getTopCoords())) {
      result.push(new GeneratorFlag(c);
      result.push(new GeneratorFlag(c.posIncrW()));
    } else {
        console.log("illegal spawnAdjacentFlags :: corner : " + c + 
          ", on flag with left coords : " + leftCoords);
    }
    return result;
  }

  // given coordinates and a segment type, returns the coords of the segment's other end 
  static getOpposingCoords(c, st) {
    if (c.equals(getLeftCoords())) {
      if (st.equals(GeneratorFlagSegmentType.NE)) {
        return c.posIncrNE();
      } else if (st.equals(GeneratorFlagSegmentType.E)) {
        return c.posIncrE();
      }
    } else if (c.equals(getRightCoords())) {
      if (st.equals(GeneratorFlagSegmentType.SE)) {
        return c.posIncrNW();
      } else if (st.equals(GeneratorFlagSegmentType.E)) {
        return c.posIncrW();
      }
    } else if (c.equals(getTopCoords())) {
      if (st.equals(GeneratorFlagSegmentType.SE)) {
        return c.posIncrSE();
      } else if (st.equals(GeneratorFlagSegmentType.NE)) {
        return c.posIncrSW();
      }
    } 

    console.log("illegal getOpposingCoords :: corner : " + c + 
      ", st : "+ st +", on flag with left coords : " + leftCoords);
    return null;

  }

  getAllCoords() {
    return [getLeftCoords(), getRightCoords(), getTopCoords()];
  }

  getLeftCoords() {
    return leftCoords;
  }

  getRightCoords() {
    return posIncrW(leftCoords);
  }

  getTopCoords() {
    return posIncrNW(leftCoords);
  }

  setSelectedSegment(selectedType) {
    selectedSegment = selectedType;
  }

  selectedCallback(selectedType) {
    if (selectedType == selectedSegment) {
      // this is a deselection 
      setSelectedSegment(null);
      this.props.posUpdateCallback(this, selectedType, false);

    } else if (selectedSegment == null) {
      setSelectedSegment(selectedType);
      this.props.posUpdateCallback(this, selectedType, true);
    }
  }

  return (
    <div>  
      <GeneratorFlagSegment segType=GeneratorFlagSegmentType.NE selectedCallback={this.props.selectedCallback}/>
      <GeneratorFlagSegment segType=GeneratorFlagSegmentType.E selectedCallback={this.props.selectedCallback}/>
      <GeneratorFlagSegment segType=GeneratorFlagSegmentType.SE selectedCallback={this.props.selectedCallback}/>
    </div>
  );
}

File: src/generator_flag.py
from enum import Enum

class GeneratorFlagSegmentType(Enum):
    NE = 1
    E = 2
    SE = 3

class GeneratorFlag:

	def __init__(self, leftCoords):
		# the coords of the left corner of the flag. other corner locations are inferred
    	self.leftCoords = leftCoords 
    	# properties set when a flag is added to the path
    	self.pathSegType = None
    	self.pathEndCoords = None

	# given a coord
	# returns [] of flags that will complete this intersection
	def spawnAdjacentFlags(self, c):
		result = [];
		#left
		result.append(GeneratorFlag(c.posIncrW()))
		result.append(GeneratorFlag(c.posIncrSW()))

		#top
		result.append(GeneratorFlag(c.posIncrNE().posIncrW()))
		result.append(GeneratorFlag(c.posIncrNE().posIncrNE()))

		#right
		result.append(GeneratorFlag(c.posIncrE()))
		result.append(GeneratorFlag(c.posIncrE().posIncrSW()))

		return result

	# given coordinates and a segment type, returns the coords of the segment's other end 
	def getOpposingCoords(self, c, st):
		if c == getLeftCoords():
		    if st.equals(GeneratorFlagSegmentType.NE):
		        return c.posIncrNE()
		    elif st.equals(GeneratorFlagSegmentType.E):
		        return c.posIncrE()
		elif c == getRightCoords():
		    if st.equals(GeneratorFlagSegmentType.SE):
		        return c.posIncrNW()
		    elif st.equals(GeneratorFlagSegmentType.E):
		        return c.posIncrW()
		elif c == getTopCoords():
		    if st.equals(GeneratorFlagSegmentType.SE):
		        return c.posIncrSE()
		    elif st.equals(GeneratorFlagSegmentType.NE):
		        return c.posIncrSW()

		print("illegal getOpposingCoords :: corner : " + c + 
            ", st : "+ st +", on flag with left coords : " + self.leftCoords)
    	return None


	def getAllCoords(self):
		return [getLeftCoords(), getRightCoords(), getTopCoords()]

	def getLeftCoords(self):
		return self.leftCoords

	def getRightCoords(self):
		return self.leftCoords.posIncrW()

	def getTopCoords(self):
		return self.leftCoords.posIncrNW()

	def setFlagAsSelected(self, pathEndCoords, pathSegType):
		self.pathSegType = pathSegType
		self.pathEndCoords = pathEndCoords

	def getPathEndCoords(self):
		return self.pathEndCoords

	def setFlagAsUnselected(self):
		self.pathSegType = None
		self.pathEndCoords = None

	def isInPlay(self):
		return self.pathEndCoords is None and self.pathSegType is None

	def __eq__(self, f):
        return isinstance(obj, f) and f.getLeftCoords() == self.leftCoords
